### PR TITLE
Integrate with Hugging Face

### LINKED ISSUE #1: Improve discoverability on HF
Hi,

Niels here from the open-source team at Hugging Face. Congrats on your work!

It would be great to make the checkpoints available on HF, along with dedicated model cards, which also enhances the visibility of your work. People could find them when searching https://huggingface.co/models?pipeline_tag=mask-generation&sort=trending.

This also comes with download stats, so that you see how many times people download your models.

e.g. facebook/sam-vit-base has been downloaded 530k times last month: https://huggingface.co/facebook/sam-vit-base.

## 1. Upload checkpoints in seperate model repos

Ideally, each checkpoint is pushed to a separate model repository. Here's a guide for uploading models to the 🤗 hub: https://huggingface.co/docs/hub/models-uploading. For this use case, I assume that the [PyTorchModelHubMixin](https://huggingface.co/docs/hub/models-uploading#upload-a-pytorch-model-using-huggingfacehub) class is best suited, it creates a config.json along with safetensors for each model.

## 2. Make the model Transformers compatible

In case you want your models to be usable through the Transformers library with `trust_remote_code=True`, I highly recommend following this guide: https://huggingface.co/docs/transformers/custom_models. It basically allows people to use your backbones using the `AutoModel` and `AutoModelForMaskGeneration` APIs.

We recently did the same with the MambaVision author as can be seen here: https://huggingface.co/collections/nvidia/mambavision-66943871a6b36c9e78b327d3.

Let me know if you need any help regarding this!

Cheers,

Niels
ML Engineer @ HF 🤗

### PR DESCRIPTION
**Pull request by NielsRogge**
This PR fixes #1. It aims to make it possible to load any SAM 2 model directly from the hub.

The checkpoints are already part of the facebook org:
* https://huggingface.co/facebook/sam2-hiera-tiny
* https://huggingface.co/facebook/sam2-hiera-small
* https://huggingface.co/facebook/sam2-hiera-base-plus
* https://huggingface.co/facebook/sam2-hiera-large.

This PR allows to do the following:

```python
from sam2.sam2_image_predictor import SAM2ImagePredictor

predictor = SAM2ImagePredictor.from_pretrained("facebook/sam2-hiera-small")
```

This also ensures that [download stats](https://huggingface.co/docs/hub/models-download-stats) work, meaning you'll be able to see how many times people actually load one of these checkpoints from the hub. This will also allow us to compare to the download stats of Transformers which will be interesting :)

To do:

- [x] update README to link to HF artifacts
- [ ] update notebooks (?)
- [x] support video predictor

### CODE REVIEW

**Comment by bhack** [comment]
Is this also in the to-do list at https://github.com/huggingface/transformers/pull/32394

File: sam2/sam2_image_predictor.py [hunk 1]
```diff
@@ -13,7 +13,7 @@
 from PIL.Image import Image
 
 from sam2.modeling.sam2_base import SAM2Base
-
+from sam2.build_sam import build_sam2_hf
```
Comment (haithamkhedr) [inline sam2/sam2_image_predictor.py:16]: Similarly, lets make this optional

File: sam2/build_sam.py [hunk 2]
```diff
@@ -11,6 +11,8 @@
 from hydra.utils import instantiate
 from omegaconf import OmegaConf
 
+from huggingface_hub import hf_hub_download
```
Comment (haithamkhedr) [inline sam2/build_sam.py:14]: Can we move this inside the function to make this an optional dependency?

**Review by haithamkhedr — CHANGES_REQUESTED** [review]
@NielsRogge Thanks for supporting ckpt download from HF. Can we make `huggingface_hub` a soft dependency?

File: README.md
Comment (haithamkhedr) [inline README.md:1]: I think it would be better to have a seperate README section for HF checkpoints paths as well as usage, leaving the original sample usage as is (since `huggingface_hub` is also an optional dependancy)

File: sam2/build_sam.py [hunk 3]
```diff
@@ -76,6 +76,30 @@ def build_sam2_video_predictor(
...
+
+    model_id_to_filenames = {
+        "facebook/sam2-hiera-tiny": ("sam2_hiera_t.yaml", "sam2_hiera_tiny.pt"),
+        "facebook/sam2-hiera-small": ("sam2_hiera_s.yaml", "sam2_hiera_small.pt"),
+        "facebook/sam2-hiera-base-plus": ("sam2_hiera_b+.yaml", "sam2_hiera_base_plus.pt"),
+        "facebook/sam2-hiera-large": ("sam2_hiera_l.yaml", "sam2_hiera_large.pt"),
+    }
+    config_name, checkpoint_name = model_id_to_filenames[model_id]
+    ckpt_path = hf_hub_download(repo_id=model_id, filename=checkpoint_name)
+    return build_sam2_video_predictor(config_file=config_name, ckpt_path=ckpt_path, **kwargs)
```
Comment (haithamkhedr) [inline sam2/build_sam.py:91]: This should be a call to `build_sam2` ?

File: sam2/build_sam.py [hunk 4]
```diff
@@ -76,6 +76,30 @@ def build_sam2_video_predictor(
...
+
+
+def build_sam2_video_predictor_hf(model_id, **kwargs):
+
+    from huggingface_hub import hf_hub_download
+
+    config_file = hf_hub_download(repo_id=model_id, filename=f"{model_id}.yaml")
```
Comment (haithamkhedr) [inline sam2/build_sam.py:98]: We also need a dictionary mapping `model_id` to config names similar to `model_id_to_filenames`

File: README.md [hunk 5]
```diff
@@ -101,6 +101,36 @@ with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
...
+
+```python
+import torch
+from sam2.sam2_video_predictor import SAM2VideoPredictor
+
+predictor = SAM2VideoPredictor.from_pretrained("facebook/sam2-hiera-large")
+
+with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
+    predictor.set_image(<your_image>)
+    masks, _, _ = predictor.predict(<input_prompts>)
```
Comment (haithamkhedr) [inline README.md:131]: Can we replace this with video predictor usage:

```python
with torch.inference_mode(), torch.autocast("cuda", dtype=torch.bfloat16):
    state = predictor.init_state(<your_video>)

    # add new prompts and instantly get the output on the same frame
    frame_idx, object_ids, masks = predictor.add_new_points(state, <your_prompts>):

    # propagate the prompts to get masklets throughout the video
    for frame_idx, object_ids, masks in predictor.propagate_in_video(state):
```

**Review by haithamkhedr — APPROVED** [review]
Thanks for adding support to load the models through HF hub.

**Comment by giswqs** [comment]
This is great! Is there a plan to make it compatible with transformers, like SAM 1?

https://huggingface.co/docs/transformers/en/tasks/mask_generation

**Comment by NielsRogge** [comment]
Yes there is, the authors are adding a PR here: https://github.com/huggingface/transformers/pull/32394

**Comment by giswqs** [comment]
Fantastic! Can't wait for the PR to be merged!

**Comment by leonmkim** [comment]
Sorry potentially dumb question, but should this [line](https://github.com/facebookresearch/segment-anything-2/blob/dce7b5446f260cef9fdf3d3f1bc81519302d386f/sam2/sam2_video_predictor.py#L124) in the video predictor be changed from `return cls(sam_model)` to `return sam_model`? It seems like the former would work for the image predictor based on its constructor but not the video predictor. But I could be wrong :p

**Comment by haithamkhedr** [comment]
> Sorry potentially dumb question, but should this [line](https://github.com/facebookresearch/segment-anything-2/blob/dce7b5446f260cef9fdf3d3f1bc81519302d386f/sam2/sam2_video_predictor.py#L124) in the video predictor be changed from `return cls(sam_model)` to `return sam_model`? It seems like the former would work for the image predictor based on its constructor but not the video predictor. But I could be wrong :p

Good catch. Fixed in #205
